Icing/http1 separation

icing opened the pull request:
This is a rather large PR that requires some explanation:

## Motivation

Historically, httpd grew on HTTP/1.0 and added later HTTP/1.1 specifics like `Expect` headers
and `Transfer-Encoding` in its request/response handling. It was assumed that HTTP/1.2, if
it ever came, would add more special cases but basically stay very similar on the wire.

With the standardization of HTTP/2 as a binary format, this held no longer true. The HTTP
IETF specifications made later a switch to differentiate between the *serialization* of the
different protocol versions and the inherent HTTP *semantics*. What a request header
*means*, irregardless of how it arrived in a server. This PR is a start in reflecting this separation inside the Apache httpd as well.

The httpd filter stack already defines the different layers necessary:

 * RESOURCE:  the resource served
 * CONTENT: content-encoding, e.g. gzip
 * PROTOCOL:  generic HTTP
 * TRANSCODE:  HTTP/1.x or HTTP/2
 * CONNECTION: TLS security
 * NETWORK:  sockets

The addition of HTTP/2 to the server made it necessary to replace the existing PROTOCOL/TRANSCODE filters for its requests. As the existing ones *knew* about HTTP/1.1 specifics that do not apply to h2. This led to code duplication of generic HTTP semantics. Improved error handling in one part did not necessarily made it into the other.

## The Change

The changes can be divided in three parts and are easier understood that way.

### Core HTTP Handling

Three new meta bucket type have been added: 

 1. `request`: holds `method`, `uri`, `protocol` and `headers` of a HTTP request.
 2. `response`: holds `status`, `reason`, `headers` and `notes` of a HTTP response.
 3.  `headers`: holds `headers` generated before the end of a HTTP request or response. Used for trailers.

With these new meta buckets, input/output filters on protocol level will see buckets in the following patterns:

```
input: (REQUEST|ERROR) (DATA|FLUSH|ERROR|META|HEADERS)* EOS
output: RESPONSE+ (DATA|FLUSH|ERROR|META|HEADERS)* EOS
```

This means that `HTTP` protocol processing is no longer concerned with parsing incoming bytes or formatting output. All `DATA` buckets on this filter level are part of request/response bodies and have no special protocol interpretation. Interim responses, such as the common 101, are sent as `response` buckets and will not confuse any content filters.

The core request function `ap_read_request()` was changed to expect either a `request` or `error` bucket from the transcode/network input filters. Transcoding input filters use the `error` bucket to trigger a HTTP error response by `ap_read_request()`.

All HTTP/1.1 parsing and formatting has been moved from core/mod_http into the new `mod_http1`.

### HTTP/1.1 Handling

The code to parse and format HTTP/1.1 messages has been moved to the new `mod_http1` (which is a mandatory module, just like mod_http, no loading necessary).

For example, in the `pre_read_request` hook, the module installs the filter that parses HTTP/1.x requests *if* the connection protocol is `http/1.1`. It also has the filters for chunking and response formatting. Moved here from `mod_http` and slightly adapted.

### HTTP/2 Handling

`mod_http2` had major code removals, due to workarounds no longer being required. It now can use the standard `ap_process_connection()` to handle requests. Interim responses need no longer parsing. Chunked encoding needs no longer be emulated for requests without a content-length header.

## Trailers

The core HTTP filters now add incoming `headers` buckets to `r->trailers_in`. They add an outgoing `headers` bucket on EOS when `r->trailers_out` is not empty.

These are always sent on a HTTP/2 connection, but on HTTP/1.1 this is only possible when output is already being chunked. How/if/when this should be triggered is to be determined should this PR be merged. 

## Proxying

`mod_http1` adds functions to write a table of headers in HTTP/1.1 format to the server API. Those are now also used in `mod_proxy` in some places. There is more potential in having common functionality, for example parsing of a HTTP/1.1 response is quite similar to requests. And yet, the current code is duplicate.

Support for HTTP trailers has been added to `mod_proxy_http`. When a `Trailers` header is present in the client request, the backend request will forward it and use `Transfer-Encoding: chunked`, even though a Content-Length might be known.

Trailers from the backend are added to the client response and, given that either chunked HTTP/1.1 or HTTP/2 are in play on the front connection, will be sent.

## Tests

A small range of tests has been added in `test/modules/http1`, mainly to verify the  trailers implementation. The trailers checks in `test/modules/http2` have also been extended.

The Apache Perl Test framework passes on my machine. Let's see what travis says.

Commit:
http vs http/1.x separation. filters are split and new mod_http1 has
been added to keep the version 1.x relevant filters.

base tests run, but proxy handling still needs work.

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
* perl framework tests pass
 * pytest tests pass

Commit:
* moved more http1 stuff into mod_http1
 * Connection header handling in mod_http1
 * connection keepalive handling in mod_http1

Commit:
* using AP_NOTE_REQUEST_BODY_INDETERMINATE in mod_http1 and mod_http2
   to signal into generic HTTP processing that a request has a body
   of indeterminate length.

Commit:
Removing some nop code in mod_http. Removing LogLevel in http2 test env.

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
*) mod_http/mod_http1: moved more Transfer-Encoding handling into mod_http1
     mod_http2: fixed handling of trailers, adding test cases

Commit:
* mod_http1: adding trailer support in chunked output encoding
  * test: add trailer test cases for http/1 and http/2

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
*) test/modules/http1: adding out trailer test through proxy

Commit:
*) mod_proxy: allow forwarding of incoming trailers to backends
     Adding test cases. Adding common function to add the last
     chunk in chunked http/1.1 encoding.

Commit:
*) mod_http1/mod_proxy:
     - exposing functions to serialize HTTP/1.1 headers values onto
       a bucket brigade and terminating headers.
     - using new functions in mod_proxy ap_proxy_create_hdrbrgd()
     - changing removal of unwanted client headers in proxy requests
       from an O(n*m) operation to O(log(n)*m).

Commit:
*) mod_http1: rename of transcode filters to response_out/request_in.

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
*) core/mod_http1: remove some http1 filters from mod_core.h again,
     making them non-public. Do some renames for clarity of purpose
     what a filter does.
     Use post_read_request hook to install http1 specific filters
     when needed, removing knowledge from generic server/protocol.c
     handling.

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
*) core/mod_http*: adding meta bucket type RESPONSE and removing
     status+reason+notes from HEADERS bucket. HEADERS is now only
     used for trailers.
     Preparing HTTP/1.x request parsing input filter, so that
     ap_read_request() gets rid of HTTP/1.1 parsing and instead
     adds on meta bucket provided by protocol specific filters.

Commit:
*) mod_http1: init potentially uninitialized local var.

Commit:
*) mod_http1: remove output filters after a 101 protocol
     switch response has been sent.

Commit:
*) core/mod_http1: fixed the ap_http_header_filter to also trigger
     response creation when only a FLUSH was sent (this caused CI
     failures in lua's websocket test).
     test: added websocket test in the modules/proxy suite that
     checks if the python `websocket` module is installed.

Commit:
allocating APLOGNO numbers from trunk.

Commit:
*) core/mod_http/mod_http1:
     - new bucket types are now REQUEST/RESPONSE/HEADERS and carry
       the meta data for the HTTP semantics.
       (rename of HEADERS to TRAILERS pending)
     - mod_http1 has a TRANSCODE input filter that reads a HTTP/1.x
       request and returns a REQUEST bucket or ERROR bucket to
       the caller.
     - ap_read_request(), the core workhorse for request handling
       is stripped of HTTP/1.x parsing and expects REQUEST/ERROR
       buckets from the protocol filters when reading a request.
     - use of REQUEST buckets in mod_http2 to be done.

Commit:
* adding APLOGNO numbers, allocated from trunk.

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
*) mod_http2: using new REQUST buckets to forward requests
     on c2 connection to standard connection processing of the
     server.

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
*) mod_http2: some small cleanups on stream input handling.

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
*) core: adding a 'pool' member to new meta bucket structures
     so that receiving handlers have an idea about the lifetime
     of included values.

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
*) mod_http2: force AP_CONN_CLOSE on secondary connections to prevent
     anyone to care about cleanup or monitoring idle timeouts (mod_reqtimeout)
     on these.

Commit:
Merge branch 'trunk' into icing/http1-separation

jfclere commented:
Would it be possible to group the commits in less commits? It is huge to review ;-)

icing commented:
I could make a PR with just the core changes, but that one would not pass our CI. If anyone thinks it'd help, let me know.

ylavic commented on server/protocol.c:
```diff
@@ -1477,9 +1306,62 @@ request_rec *ap_read_request(conn_rec *conn)
-    /* Get the request... */
-    if (!read_request_line(r, tmp_bb) || !ap_parse_request_line(r)) {
+    r->request_time = apr_time_now();
+    rv = ap_get_brigade(r->proto_input_filters, tmp_bb, AP_MODE_READBYTES, APR_BLOCK_READ, 0);
```
IIUC, before we had proto_input_filters that were called by read_request_line() and ap_get_mime_headers_core() for each line with AP_MODE_GETLINE, while now those filters will be called once with AP_MODE_READBYTES (of zero bytes) and will get a REQUEST meta bucket from downstream/TRANSCODE filters.

I wonder how much this can break existing protocol input filters, those that still want to do something on a per header basis will now have to work/iterate on ((ap_bucket_request *)b->data)->headers_in at AP_FTYPE_PROTOCOL level still or register at AP_FTYPE_TRANSCODE + something.

Can't we leave http1_request_in_filter() as AP_FTYPE_PROTOCOL (like h2_c2_filter_request_in() is, it seems)?

icing replied:
We can make it a PROTOCOL filter, yes. Any other filter that wants to mess with the raw HTTP/1.1 bytes would need to install itself before. But maybe we add http1_request_in_filter() as REALLY_LAST in the pre_read hook?

icing replied:
I made http1_request_in_filter a PROTOCOL one.

ylavic commented on changes-entries/http1_separation.txt:
```diff
@@ -0,0 +1,31 @@
+       generic HTTP handling into mod_http1.
+       A new note AP_NOTE_REQUEST_BODY_INDETERMINATE has been
+       introduced that, when present in r->notes, indicates that
+       the length of a request body is not known.
```
Maybe a new request_rec `body_indeterminate` field/flag rather than a `notes`?
(notes are not very effectively accessed, so it depends on how often it needs to be known by filters/handlers)

icing replied:
Agreed. I started the note early when I did not want to mess with request_rec.

icing replied:
Added a field to `request_rec` named `body_indeterminate` and removed all occurences of `AP_NOTE_REQUEST_BODY_INDETERMINATE`.

ylavic commented on include/http_protocol.h:
```diff
@@ -67,6 +67,31 @@ AP_DECLARE(request_rec *) ap_create_request(conn_rec *c);
+/**
+ * Tokenize a HTTP/x.x request line.
+ * @param r the request for which to parse the line
+ * @param line the line to tokenize, will be written to.
```
The returned `line` is normalized? garbage?
Maybe a const char *line here could be copied then tokenized?
(for instance it seems that ap_parse_request_line() will overwrite r->the_request)

icing replied:
This method was, I think, static before and is now exposed. So, it would make sense to give it a cleaner calling convention at the cost of an additional pstrdup().

icing replied:
Actually, it has become purely internal to mod_http1 now. So, I'll remove it from the API and use it internally only.

icing replied:
Updated PR with change.

ylavic commented on modules/http/http_filters.c:
```diff
@@ -420,224 +94,92 @@ apr_status_t ap_http_filter(ap_filter_t *f, apr_bucket_brigade *b,
+        {
+            next = APR_BUCKET_NEXT(e);
+            if (!APR_BUCKET_IS_METADATA(e)) {
+                ctx->seen_data = (e->length != 0);
```
This may overwrite seen_data, maybe:
```
if (e->length != 0) {
    ctx->seen_data = 1;
}
```
?

ylavic replied:
or `|=`..

icing replied:
Updated PR with change.

ylavic commented on modules/http1/chunk_filter.c:
```diff
@@ -0,0 +1,762 @@
+/* Licensed to the Apache Software Foundation (ASF) under one or more
```
Could this http1/chunk_filter.c be moved from the old one in http/ (using git mv for now and probably svn move later..)?
It'd help reviewing only the changes here, and would preserve history too.

ylavic replied:
Likewise for http1/http_filters.c, it seems that all the code could be moved from http/http_filters.c or the few things remaining there don't belng in http1/ ?

icing replied:
Any suggestion how to do that now in this PR?

icing replied:
Giving up after several attempts. Changes to a git mv file show up as delete/new here. :(

ylavic commented on modules/http1/chunk_filter.c:
```diff
@@ -0,0 +1,762 @@
+    /* sanity check in case we're read twice */
+    if (ctx->at_eos) {
+        ap_log_rerror(APLOG_MARK, APLOG_TRACE3, 0, f->r,
+                      "ap_http1_transcode_in_filter: return eos");
```
s/ap_http1_transcode_in_filter/http1_body_in_filter/ ?

(there are other references to ap_http1_transcode_in_filter elsewhere which probably need to be changed too)

icing replied:
Updated PR with change.

ylavic commented on modules/http1/http1_core.c:
```diff
@@ -0,0 +1,142 @@
+static int http1_post_read_request_early(request_rec *r)
+{
+    if (!r->main && !r->prev
+        && !strcmp(AP_PROTOCOL_HTTP1, ap_get_protocol(r->connection))) {
```
r->proto_num >= HTTP_VERSION(1,1)?

icing replied:
Implemented.

ylavic commented on modules/http1/http1_core.c:
```diff
@@ -0,0 +1,142 @@
+
+    if (!r->main && !r->prev
+        && !strcmp(AP_PROTOCOL_HTTP1, ap_get_protocol(r->connection))) {
+        if (r->proto_num >= HTTP_VERSION(1,0)) {
```
r->proto_num >= HTTP_VERSION(1,1) (instead of strcmp + proto_num checks)?

icing replied:
Implemented.

ylavic commented on modules/http1/http1_filters.c:
```diff
@@ -0,0 +1,373 @@
+                ap_bucket_response *resp = e->data;
+
+                ap_log_rerror(APLOG_MARK, APLOG_TRACE2, 0, r,
+                              "ap_http1_transcode_out_filter seeing response bucket status=%d",
```
ap_http1_response_out_filter (found elsewhere in the PR too).

icing replied:
Changed.

ylavic commented on server/protocol.c:
```diff
@@ -681,245 +681,102 @@ static int field_name_len(const char *field)
+    ra_badprotocol,
+    ra_reject09,
+    ra_badmethod09,
+} ra_error;
```
`ra_`?? Maybe rl_error_ (for request_line) or http_error_?

icing replied:
Changed to `http_error` prefix.

ylavic commented on server/protocol.c:
```diff
@@ -681,245 +681,102 @@ static int field_name_len(const char *field)
-                      ap_escape_logitem(r->pool, r->the_request));
-    }
-
-    return 1;
```
`switch (error)` below? This will better catch potential unhandled cases with modern compilers.

icing replied:
I moved the new functions to `protocol_new.c` so that they do not show up as changes to removed code here.

ylavic commented:
First pass, I will come back later to it (no more free time for now, and a need to digest all the changes a bit ;)

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
Updates after review by ylavic:
- removed ap_tokenize_request_line() from the API again,
  only used internal in mod_http1.
- made ap_tokenize_request_line, which is used via the public
  ap_parse_request_line() work on a copy and not modify
  r->the_request any longer.
- ap_http_filter: fixed the check for data before headers.
- renamed all occurences of 'ap_http1_transcode_in_filter' to
  'http1_body_in_filter', leftovers from an earlier rename.

Commit:
*) core, mod_http1, mod_http2: adding field `body_indeterminate`
     to `request_rec` and using it in protocol handlers when length
     of a request body and only be determined by reading it.
     (replaces former note `AP_NOTE_REQUEST_BODY_INDETERMINATE`)

Commit:
*) mod_http1: in post_read_req hooks, check the protocol version
     of the request instead of connection.

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
*) core, mod_http1, mod_http: moved the handling of the standard
     response headers `Date` and `Server` from mod_http1 into the
     generic HTTP protocol handling.
     Response buckets not always carry those headers (values preserved
     from proxied responses), irregardless of the HTTP protocol
     versions involved.
     mod_http1: the serialization of response header into HTTP/1.x
     format always writes `Date` and `Server` first if present. This
     assured backward compatibility with clients who are accustomed
     to this order.

icing commented:
I just made a PR with only the API changes from #291.

https://github.com/icing/httpd/pull/1

Since this will never be merged to apache/httpd, I made
it against my own fork.

I started a review and commented on the new things added.
Should you find this useful, I'll make other partial PRs
for core/http1/http2/proxy changes.

Kind Regards,
Stefan

> Am 28.03.2022 um 14:16 schrieb ylavic ***@***.***>:
> 
> 
> @ylavic commented on this pull request.
> 
> In modules/http1/chunk_filter.c:
> 
> > @@ -0,0 +1,762 @@
> +/* Licensed to the Apache Software Foundation (ASF) under one or more
> 
> Likewise for http1/http_filters.c, it seems that all the code could be moved from http/http_filters.c or the few things remaining there don't belng in http1/ ?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
>

Commit:
Merge branch 'trunk' into icing/http1-separation

Commit:
PR readability:
 - splitting parts of server/protocol.c into another file, to be
   joined later. Just so diff is less confusing.
 - hopedfully correctly restoring, moving and modifying
   modules/http/chunk_filter.c in order to keep the
   history and show only the real changes done.

Commit:
PR readability: trying the get chunk_filter.c to be recognized
as moved.

Commit:
moving the file

Commit:
reapplying changes to moved chunk_filter.c

Commit:
*) mod_http1: make http1_request_in_filter a PROTOCOL filter, added last
     to the input, as suggested by ylavic on review.

Commit:
*) mod_http1: replaced occurences of previous name with
     ap_http1_response_out_filter.

icing commented:
Replaced by staged PRs #306, #307 and #311.
